consul-lambda-registrator/consul-lambda-extension/main.go

```diff
@@ -0,0 +1,125 @@
+	"github.com/aws/aws-sdk-go-v2/aws/retry"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/hashicorp/go-hclog"
+	"github.com/kelseyhightower/envconfig"
+
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/client"
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/extension"
+	"github.com/hashicorp/terraform-aws-consul-lambda-registrator/consul-lambda-registrator/trace"
+)
+
+const (
+	defaultLogLevel = "info"
+	extensionName   = "consul-lambda-extension"
+)
+
+func main() {
+	logger := hclog.New(&hclog.LoggerOptions{
+		Level: hclog.LevelFromString(getEnvOrDefault("LOG_LEVEL", defaultLogLevel)),
+	})
+
+	trace.SetLogger(trace.HCLog{Logger: logger})
+	trace.Enabled(strings.ToLower(getEnvOrDefault("TRACE_ENABLED", "false")) == "true")
+
+	err := realMain(logger.Named(extensionName))
+	if err != nil {
+		logger.Error("fatal error, exiting", "error", err)
+		os.Exit(1)
+	}
+}
+
+func realMain(logger hclog.Logger) error {
+	trace.Enter()
+	defer trace.Exit()
+
+	cfg, err := configure()
+	if err != nil {
+		return err
+	}
+
+	cfg.Logger = logger
+	ext := extension.NewExtension(cfg)
+
+	// Handle interrupts and shutdown notification
+	ctx, cancel := context.WithCancel(context.Background())
+	shutdownChannel := make(chan struct{})
+	go func() {
+		interruptChannel := make(chan os.Signal, 1)
+		signal.Notify(interruptChannel, syscall.SIGTERM, syscall.SIGINT)
+
+		select {
+		case s := <-interruptChannel:
+			logger.Info("Received signal, exiting", "signal", s)
+		case <-shutdownChannel:
+			logger.Info("Received shutdown event, exiting")
+		}
+		// Cancel our context so that all servers and go-routines exit gracefully.
+		cancel()
+	}()
+
+	err = ext.Serve(ctx)
+	if err != nil {
+		logger.Error("processing failed with an error", "error", err)
+	}
+
+	// Signal that it's time to shutdown when the extension returns.
+	shutdownChannel <- struct{}{}
+
+	return err
+}
+
+func configure() (*extension.Config, error) {
+	trace.Enter()
+	defer trace.Exit()
+
+	// Load the configuration from the environment.
+	cfg := &extension.Config{}
+	err := envconfig.Process("consul", cfg)
+	if err != nil {
+		return cfg, fmt.Errorf("failed to load configuration from environment: %w", err)
+	}
+
+	if cfg.ServiceName == "" {
+		// If the service name wasn't explicitly configured then default to the Lambda function's name.
+		cfg.ServiceName = os.Getenv("AWS_LAMBDA_FUNCTION_NAME")
+	}
+
+	sdkConfig, err := config.LoadDefaultConfig(context.Background(), config.WithRetryer(func() aws.Retryer {
+		// Adaptive mode should retry on hitting rate limits.
+		return retry.AddWithMaxBackoffDelay(retry.NewAdaptiveMode(), 3*time.Second)
+	}))
+	if err != nil {
+		return cfg, fmt.Errorf("failed to create AWS SDK configuration: %w", err)
+	}
+
+	ssmClient := client.NewSSM(&sdkConfig)
+
+	lambdaClient := client.NewLambda(&sdkConfig)
+	err = lambdaClient.Register(context.Background(), extensionName)
```

Comment:
Totally agree :+1:

Reply:
I'll make this change on the `cthain/lambda-to-mesh-support` branch.

Reply:
After looking at this again I'm not sure that this suggestion makes sense now that the Lambda clients aren't in the `client` package any more.. and only the registrator's Lambda client needs the `sdkConfig`. I think it makes the most sense to create the `sdkConfig` once on init and reuse it where applicable. Otherwise I think we need global var in the `client` package to hold it so that we don't create it more than once.

Maybe I'm missing a more elegant solution..